src/push-notifications.js

```diff
@@ -1,17 +1,140 @@
-export default function PushNotifications(config) {
-  if (config === null || config === undefined) {
-    throw new Error('Push Notifications config object is required');
+import doRequest from './doRequest';
+
+export function init(config) {
+  return new Client(config);
+}
+
+export class Client {
+  constructor(config) {
+    if (!config) {
+      throw new Error('Config object required');
+    }
+    const { instanceId, endpointOverride = null } = config;
+
+    if (instanceId === undefined) {
+      throw new Error('Instance ID is required');
+    }
+    if (typeof instanceId !== 'string') {
+      throw new Error('Instance ID must be a string');
+    }
+    if (instanceId.length === 0) {
+      throw new Error('Instance ID cannot be empty');
+    }
+
+    this.instanceId = instanceId;
+    this._endpoint = endpointOverride; // Internal only
+
+    this._initDb("beams");
+  }
+
+  get _baseURL() {
+    if (this._endpoint !== null) {
+      return this._endpoint;
+    }
+    return `https://${this.instanceId}.pushnotifications.pusher.com`;
+  }
+
+  async start() {
+    const { vapidPublicKey: publicKey } = await this._getPublicKey();
+
+    // register with pushManager, get endpoint etc
+    const token = await this._getPushToken(publicKey);
+
+    // get device id from errol
+    const response = await this._registerDevice(token);
+    // // put response.id in indexedDB
+    this.deviceId = response;
+
+    this._save(this.instanceId, token, this.deviceId);
+  }
+
+  async _getPublicKey() {
+    const path = `${this._baseURL}/device_api/v1/instances/${encodeURIComponent(
+      this.instanceId
+    )}/web-vapid-public-key`;
+    return doRequest('GET', path);
   }
-  function start() {
-    console.log('starting');
+
+  async _getPushToken(publicKey) {
+    let sub;
+    try {
+      window.navigator.serviceWorker.register('sw.js');
+      const reg = await window.navigator.serviceWorker.ready;
+      sub = await reg.pushManager.subscribe({
+        userVisibleOnly: true,
+        applicationServerKey: urlBase64ToUInt8Array(publicKey),
+      });
+    } catch (e) {
+      return Promise.reject(e);
+    }
+
+    return btoa(JSON.stringify(sub));
+  }
+
+  async _registerDevice(token) {
+    const path = `${this._baseURL}/device_api/v1/instances/${encodeURIComponent(
+      this.instanceId
+    )}/devices/web`;
+
+    return doRequest('POST', path, { token });
   }
 
-  function setUserId() {
-    console.log('hey');
+  _initDb(dbName) {
+    var request = indexedDB.open(dbName);
+
+    request.onerror = function(event) {
+      console.log(`Database error: " + ${event.target.errorCode}`);
```

Comment:
you can use `console.error`, which is the same as `console.log` but it will appear in red in the console